extensions/amp-truncate-text/1.0/truncate-text.js

```diff
@@ -0,0 +1,152 @@
+import {setStyle} from '../../../src/style';
+import {truncateText} from './truncation';
+import {
+  useCallback,
+  useEffect,
+  useImperativeHandle,
+  useLayoutEffect,
+  useRef,
+  useState,
+} from '../../../src/preact';
+import {useStyles} from './truncate-text.jss';
+
+/**
+ * @param {!TruncateTextDef.Props} props
+ * @param {{current: (!TruncateTextDef.TruncateTextApi|null)}} ref
+ * @return {PreactDef.Renderable}
+ */
+function TruncateTextWithRef(
+  {
+    slotPersistent,
+    slotCollapsed,
+    slotExpanded,
+    onBeforeExpand,
+    onAfterCollapse,
+    ...rest
+  },
+  ref
+) {
+  const wrapperRef = useRef();
+  const collapsedRef = useRef();
+  const persistentRef = useRef();
+
+  const onBeforeExpandRef = useRef(onBeforeExpand);
+  const onAfterCollapseRef = useRef(onAfterCollapse);
+
+  const [expanded, setExpanded] = useState(false);
+  const [ready, setReady] = useState(false);
+
+  // Provide API actions
+  useImperativeHandle(ref, () => ({
+    expand: () => setExpanded(true),
+    collapse: () => setExpanded(false),
+  }));
+
+  // TODO(rcebulko): Rewrite `truncateText` for Preact
+  /** Perform truncation on contents. */
+  const truncate = useCallback(() => {
+    truncateText({
+      container: wrapperRef.current,
+      overflowNodes: [persistentRef.current, collapsedRef.current],
+    });
+  }, []);
+
+  // Truncate the text when expanded/collapsed
+  useLayoutEffect(() => {
+    // Update container element attributes
+    if (expanded) {
+      onBeforeExpandRef.current && onBeforeExpandRef.current();
+    } else {
+      onAfterCollapseRef.current && onAfterCollapseRef.current();
+    }
+
+    // Truncate contents to fit/expand container
+    truncate();
+  }, [expanded, truncate]);
+
+  // After the first layout, measure/truncate/display
+  useEffect(() => {
+    truncate();
+    setReady(true);
+  }, [truncate]);
+
+  // Don't display contents until after the first measure/truncate
+  useLayoutEffect(() => {
+    setStyle(wrapperRef.current, 'visibility', ready ? 'visible' : 'hidden');
+  }, [ready]);
+
+  const classes = useStyles();
+  const slots = {
+    'default': <slot />,
+    'persistent': slotPersistent,
+    'collapsed': slotCollapsed,
+    'expanded': slotExpanded,
+  };
+
+  const TruncateSlot = forwardRef(({name, className, ...rest}, ref) => (
+    <span
+      className={`i-amphtml-truncate-${name}-slot ${className || ''}`}
+      ref={ref}
+      {...rest}
+    >
+      {slots[name]}
+    </span>
+  ));
+
+  return (
+    <Wrapper
+      ref={wrapperRef}
+      // Required to un-set defaults that break truncation
+      wrapperStyle={{position: null}}
+      wrapperClassName={`i-amphtml-truncate-content ${
+        classes.truncateTextContent
+      } ${
+        expanded
+          ? classes.truncateTextExpandedContent
+          : classes.truncateTextCollapsedContent
+      }`}
+      {...rest}
+    >
+      <TruncateSlot name="default" />
```

Comment:
I still think calling these slots is specific to AMP/Bento mode. For instance, in Preact mode, the user could use it this way:
```
<TruncateText collapsed={<span>See more</span>}>
  lots of text here
</TruncateText>
```

Reply:
Helper ended up removed in local branch; will pushoon